features/steps/steps.py

```diff
@@ -1,33 +1,294 @@
+    return save_doi_to_temporary_file(o_doi_label)
+
+def reserve_action_run(node_value,input_value):
+    # Helper function to 'reserve' a given input_value.
+    global g_action_expectant_bad
+    logger.info(f"when node_value,input_value {node_value,input_value}")
+    o_doi_label = None
+
+    db_name = 'doi_temp.db'
+    _action = DOICoreActionReserve(db_name=db_name)
+
+    o_doi_label = _action.run(
+                      input=input_value,
+                      node=node_value, submitter='my_user@my_node.gov',
+                      dry_run=True,force=True)
+
+    return save_doi_to_temporary_file(o_doi_label)
+    #return o_doi_label
+
+def draft_output_compare(output_file, ref_output_value):
+    # Function compare two XML files created from 'draft' or 'reserve' actions.
+    # Assumption(s): 
+    #   1.  The name of the new XML file is defined in get_temporary_output_filename().
+    #   2.  The name of the historical name is ref_output_value
+    logger.info(f"output_file,ref_output_value {output_file},{ref_output_value}")
+
+    o_fields_differ_list, o_values_differ_list, o_record_index_differ_list = DOIDiffer.doi_xml_differ(ref_output_value,
+                                                                                                      output_file)
+
+
+    logger.info(f'different fields are {o_fields_differ_list}')
+
+    assert len(o_fields_differ_list) is 0
+
+    return 1
 
-@given('a valid PDS4 {input_type} at url {input_value}')
-def given_valid_input(context, input_type, input_value):
-    logger.info(f"given {input_type} {input_value}")
+def reserve_output_compare(output_file, ref_output_value):
+    logger.info(f"output_file,ref_output_value {output_file},{ref_output_value}")
+    # Use the same function draft_output_compare() to compare a DOI (XML output) from the 'reserve' action.
+    draft_output_compare(output_file,ref_output_value)
 
+def initialize_normal_draft():
+    global g_action_expectant_bad, g_draft_condition
+    # Set to g_action_expectant_bad to signify to not expect the action to be bad.
+    g_action_expectant_bad = False
+    g_draft_condition = TestCondition.NORMAL
 
-@given('an invalid PDS4 {input_type} at {input_value}')
+def initialize_normal_reserve():
+    global g_action_expectant_bad, g_reserve_condition
+    # Set to g_action_expectant_bad to signify to not expect the action to be bad.
+    g_action_expectant_bad = False
+    g_reserve_condition = TestCondition.NORMAL
+
+@given('a valid PDS4 label at {input_value}')
+def given_valid_draft_input(context, input_value):
+    global g_action_expectant_bad, g_draft_condition
+    logger.info(f"given {input_value}")
+    initialize_normal_draft()
+    context.input_value = input_value
+    logger.info(f"g_action_expectant_bad {g_action_expectant_bad}")
+
+@given('a valid reserve PDS4 label at input_type,input_value {input_type},{input_value}')
+def given_valid_reserve_input(context, input_type, input_value):
```

Comment:
Since these functions are shared between 'reserve' and 'draft', we have made a generic function to set the input_value:

@given('a valid PDS4 label at {input_value}')
def given_valid_action_input(context, input_value):
    logger.info(f"given {input_value}")
    context.input_value = input_value  # Don't forget to set the input_value in context to be available for other functions.

Reply:
Perfect